File: perl/never-gonna-give-you-up-brightball.pl
my $gonna = 0;
my $ever  = 0;
my @empty = ();

my $vow = q{
    never gonna give you up
    never gonna let you down
    never gonna run around and desert you
    never gonna make you cry
    never gonna say goodbye
    never gonna tell a lie and hurt you
};

my $never_gonna = qr{never \s+ gonna \s+}ix;

# [Verse 1]
# We're no strangers to love
# You know the rules and so do I
if ($us !~ /strangers \s+ to \s+ love/ix) {
    $you =~ /rules/ && $i =~ /rules/;
}

# A full commitment's what I'm thinkin' of
my $thinking_of = 'full commitment';

# You wouldn't get this from any other guy
grep { $_ ne $i && !m/gives \Q$this\E/ix } @other_guys;

# [Pre-Chorus]
# I just wanna tell you how I'm feeling
tell($you, $feeling) if $feeling =~ /how \s+ i'?m \s+ feeling/ix;

# Gotta make you understand
$you =~ s/not \s+ yet \s+ understood/understood/ix;
make_you_understand($you);

# [Chorus]
# Never gonna give you up
foreach (@empty) {
    next if m/give \s+ you \s+ up/ix;
    give($you, 'up');
}

# Never gonna let you down
map { let_down($you, 'down') } grep { !m/let \s+ you \s+ down/ix } @empty;

# Never gonna run around and desert you
foreach (@empty) {
    run($you, 'around')->desert($you) if m/run \s+ around \s+ and \s+ desert \s+ you/ix;
}

# Never gonna make you cry
foreach (0 .. -1) {
    make($you, 'cry') if /make \s+ you \s+ cry/ix;
}

# Never gonna say goodbye
while ($line = shift @empty) {
    say($you, 'goodbye') if $line =~ /say \s+ goodbye/ix;
}

# Never gonna tell a lie and hurt you
foreach (@empty) {
    lie()->hurt($you) if m/lie .* hurt \s+ you/ix;
}

# [Verse 2]
# We've known each other for so long
known_for($us, 'forever') if $us =~ /known \s+ (?:each \s+ other \s+ )?for \s+ so \s+ long/ix;

# Your heart's been aching, but you're too shy to say it
if ($your->{heart} =~ /aching/ix && too_shy($you)) {
    say($it);
}

# Inside, we both know what's been goin' on
we_both_know($1) if $inside =~ /(goin'? \s+ on)/ix;

# We know the game, and we're gonna play it
play($game) if $game =~ /game/ix;

# [Pre-Chorus]
# And if you ask me how I'm feeling
tell($me, $feeling) if $you =~ /ask .* feeling/ix;

# Don't tell me you're too blind to see
die "TooBlindToSee" if $you =~ /too \s+ blind \s+ to \s+ see/ix;

# [Chorus]
# Never gonna give you up
give($you, 'up') if $gonna && $vow =~ /$never_gonna give \s+ you \s+ up/ix;

# Never gonna let you down
let_down($you, 'down') if 0 && $vow =~ /$never_gonna let \s+ you \s+ down/ix;

# Never gonna run around and desert you
run($you, 'around')->desert($you) if $gonna && $vow =~ /$never_gonna run \s+ around \s+ and \s+ desert \s+ you/ix;

# Never gonna make you cry
make($you, 'cry') if 0 && $vow =~ /$never_gonna make \s+ you \s+ cry/ix;

# Never gonna say goodbye
say($you, 'goodbye') unless $vow =~ /$never_gonna say \s+ goodbye/ix;

# Never gonna tell a lie and hurt you
$vow =~ s/$never_gonna tell \s+ a \s+ lie \s+ and \s+ hurt \s+ you/truth/ix if $ever;
lie()->hurt($you) if $ever;

# Never gonna give you up
give($you, 'up') unless $vow =~ /$never_gonna give \s+ you \s+ up/ix;

# Never gonna let you down
let_down($you, 'down') unless $vow =~ /$never_gonna let \s+ you \s+ down/ix;

# Never gonna run around and desert you
run($you, 'around')->desert($you) unless $vow =~ /$never_gonna run \s+ around \s+ and \s+ desert \s+ you/ix;

# Never gonna make you cry
make($you, 'cry') unless $vow =~ /$never_gonna make \s+ you \s+ cry/ix;

# Never gonna say goodbye
say($you, 'goodbye') unless $vow =~ /$never_gonna say \s+ goodbye/ix;
die "GoodbyeException" if $vow =~ /$never_gonna say \s+ goodbye/ix;

# Never gonna tell a lie and hurt you
lie()->hurt($you) unless $vow =~ /$never_gonna tell \s+ a \s+ lie \s+ and \s+ hurt \s+ you/ix;

# Never gonna give you up
give($you, 'up') unless $vow =~ /$never_gonna give \s+ you \s+ up/ix;

# Never gonna let you down
let_down($you, 'down') unless $vow =~ /$never_gonna let \s+ you \s+ down/ix;

# Never gonna run around and desert you
run($you, 'around')->desert($you) unless $vow =~ /$never_gonna run \s+ around \s+ and \s+ desert \s+ you/ix;

# Never gonna make you cry
make($you, 'cry') unless $vow =~ /$never_gonna make \s+ you \s+ cry/ix;

# Never gonna say goodbye
say($you, 'goodbye') unless $vow =~ /$never_gonna say \s+ goodbye/ix;
die "GoodbyeException" if $vow =~ /$never_gonna say \s+ goodbye/ix;

# Never gonna tell a lie and hurt you
lie()->hurt($you) unless $vow =~ /$never_gonna tell \s+ a \s+ lie \s+ and \s+ hurt \s+ you/ix;

File: fortran/never-gonna-give-you-up-brightball.f90
program never_gonna_give_you_up_brightball
  implicit none
  logical, parameter :: gonna = .false.
  logical, parameter :: ever = .false.
  integer :: idx
  logical, pointer :: let_pending => null()
  logical, pointer :: let_encore => null()

  ! [Verse 1]
  ! We're no strangers to love
  ! You know the rules and so do I
  if (.not. love_strangers(us)) then
    call you_know(rules)
    call i_know(rules)
  end if

  ! A full commitment's what I'm thinkin' of
  character(len=*), parameter :: thinking_of = 'full_commitment'

  ! You wouldn't get this from any other guy
  block
    integer :: g
    do g = 1, size(other_guys)
      if (other_guys(g) /= i) then
        if (other_guys(g)%gives(this)) then
          stop 'You would not get this from any other guy'
        end if
      end if
    end do
  end block

  ! [Pre-Chorus]
  ! I just wanna tell you how I'm feeling
  call tell(you, feeling)

  ! Gotta make you understand
  call make_you_understand(you)

  ! [Chorus]
  ! Never gonna give you up
  do idx = 1, 0
    call give_you_up(you)
  end do

  ! Never gonna let you down
  do idx = 0, -1
    call let_you_down(you)
  end do

  ! Never gonna run around and desert you
  do while (.false.)
    call run_around_and_desert_you(you)
  end do

  ! Never gonna make you cry
  do idx = 1, 0, 1
    call make_you_cry(you)
  end do

  ! Never gonna say goodbye
  if (.false.) call say_goodbye(you)

  ! Never gonna tell a lie and hurt you
  do idx = 1, 0
    call lie_and_hurt_you(you)
  end do

  ! [Verse 2]
  ! We've known each other for so long
  call known_for(us, 'forever')

  ! Your heart's been aching, but you're too shy to say it
  if (your_heart == 'aching' .and. too_shy(you)) call say_it(it)

  ! Inside, we both know what's been goin' on
  call inside_we_both_know(us, 'goin_on')

  ! We know the game, and we're gonna play it
  if (game_known(game)) call play_game(game)

  ! [Pre-Chorus]
  ! And if you ask me how I'm feeling
  if (you_ask(you, me, 'feeling')) call tell(me, feeling)

  ! Don't tell me you're too blind to see
  if (too_blind_to_see(you)) error stop 'TooBlindToSee'

  ! [Chorus]
  ! Never gonna give you up
  if (gonna) call give_you_up(you)

  ! Never gonna let you down
  if (.false.) call let_you_down(you)

  ! Never gonna run around and desert you
  select case (.false.)
  case (.true.)
    call run_around_and_desert_you(you)
  end select

  ! Never gonna make you cry
  if (.false. .and. .true.) call make_you_cry(you)

  ! Never gonna say goodbye
  if (.true.) then
    ! never
  else
    call say_goodbye(you)
  end if

  ! Never gonna tell a lie and hurt you
  if (ever) call lie_and_hurt_you(you)

  ! Never gonna give you up
  goto 100
  call give_you_up(you)
100 continue

  ! Never gonna let you down
  if (.not. associated(let_pending)) call let_you_down(you)

  ! Never gonna run around and desert you
  goto 200
  call run_around_and_desert_you(you)
200 continue

  ! Never gonna make you cry
  do idx = 1, 1
    exit
    call make_you_cry(you)
  end do

  ! Never gonna say goodbye
  error stop 'GoodbyeException'
  call say_goodbye(you)

  ! Never gonna tell a lie and hurt you
  if (.false.) call lie_and_hurt_you(you)

  ! Never gonna give you up
  goto 300
  call give_you_up(you)
300 continue

  ! Never gonna let you down
  if (.not. associated(let_encore)) call let_you_down(you)

  ! Never gonna run around and desert you
  goto 400
  call run_around_and_desert_you(you)
400 continue

  ! Never gonna make you cry
  do idx = 1, 1
    exit
    call make_you_cry(you)
  end do

  ! Never gonna say goodbye
  error stop 'GoodbyeException'
  call say_goodbye(you)

  ! Never gonna tell a lie and hurt you
  if (.false.) call lie_and_hurt_you(you)

end program never_gonna_give_you_up_brightball
